cli/flox-core/src/activations.rs

```diff
@@ -266,6 +266,24 @@ impl Activation {
         self.attached_pids
             .retain(|attached_pid| attached_pid.pid != pid);
     }
+
+    /// Remove any PIDs that are not running, except if they have an expiration in the future.
+    ///
+    /// Returns a boolean indicating whether any PIDs were removed.
+    pub fn remove_terminated_pids(&mut self) -> bool {
+        let now = OffsetDateTime::now_utc();
+        let old_len = self.attached_pids.len();
+        self.attached_pids.retain(|attached_pid| {
+            if let Some(expiration) = attached_pid.expiration {
+                // If the PID has an unreached expiration, retain it even if it
+                // isn't running
+                now < expiration || pid_is_running(attached_pid.pid)
```

Comment:
Q: why would the expiration be in the future, and why do we retain it if its not running?

Reply:
I don't recall the full flow and I can't find the diagram from the activate/watchdog rewrite but it's used for in-place activations where we register an intent to attach in one shell and then it's replaced by the evaling shell: https://github.com/flox/flox/blob/81100746f1c223933b7fd05a7ad0a997d7dc07e3/assets/environment-interpreter/activate/activate.d/attach-inplace.bash#L7-L23

(the logic is a straight copy of what was there before)